src/flatcollections-array/FlatArray/FlatArray.T/FlatArray.Functional.Filter.cs

```diff
@@ -0,0 +1,84 @@
+namespace System;
+
+partial struct FlatArray<T>
+{
+    public FlatArray<T> Filter(Predicate<T> predicate)
+    {
+        _ = predicate ?? throw new ArgumentNullException(nameof(predicate));
+
+        if (length == default)
+        {
+            return default;
+        }
+
+        var startLength = length < InnerAllocHelper.DefaultPositiveCapacity ? length : InnerAllocHelper.DefaultPositiveCapacity;
+
+        var resultItems = new T[startLength];
+        var resultLength = 0;
+
+        for (int i = 0; i < length; i++)
```

Comment:
If actual `main` is merged into this one, the cycle can be replaced by `InternalForEach`:
```cs
        InternalForEach(item =>
        {
            if (predicate.Invoke(item) is false)
            {
                return;
            }

            if (resultLength == resultItems.Length)
            {
                InnerBufferHelper.EnlargeBuffer(ref resultItems);
            }

            resultItems[resultLength++] = item;
        });
```